[BUG] Cannot build windows static library with embedded source

`slang-bootstrap` relies symbols from `slang.dll`. When you build it as a static library, on linux it somehow resolves those symbols, but on windows it cannot resolve them. (I say somehow, because they are different executables (my app and `slang-bootstrap`). I'd like an explanation, if anyone knows how)

Error:
```
20>------ Build started: Project: slang-bootstrap, Configuration: Debug x64 ------
20>main.obj : error LNK2001: unresolved external symbol __imp_spCompile
20>main.obj : error LNK2001: unresolved external symbol __imp_spSetDiagnosticCallback
20>main.obj : error LNK2001: unresolved external symbol __imp_slang_shutdown
20>main.obj : error LNK2001: unresolved external symbol "__declspec(dllimport) void __cdecl spSetCommandLineCompilerMode(struct slang::ICompileRequest *)" (__imp_?spSetCommandLineCompilerMode@@YAXPEAUICompileRequest@slang@@@Z)
20>main.obj : error LNK2001: unresolved external symbol __imp_spCreateCompileRequest
20>main.obj : error LNK2001: unresolved external symbol __imp_slang_createGlobalSession
20>main.obj : error LNK2001: unresolved external symbol __imp_spDestroyCompileRequest
20>main.obj : error LNK2001: unresolved external symbol __imp_spProcessCommandLineArguments
20>main.obj : error LNK2001: unresolved external symbol __imp_slang_createGlobalSessionWithoutCoreModule
20>D:\DEV\Github\PK_Gen\.PROJ\generators\Debug\bin\slang-bootstrap.exe : fatal error LNK1120: 9 unresolved externals
20>Done building project "slang-bootstrap.vcxproj" -- FAILED.
21>------ Build started: Project: slang-embedded-core-module, Configuration: Debug x64 ------
21>Generating slang-core-module-generated.h
21>'..\..\..\..\generators\Debug\bin\slang-bootstrap.exe' is not recognized as an internal or external command,
21>operable program or batch file.
21>C:\Program Files\Microsoft Visual Studio\2022\Community\MSBuild\Microsoft\VC\v170\Microsoft.CppCommon.targets(254,5): error MSB8066: Custom build for 'D:\DEV\Github\PK_Gen\.PROJ\CMakeFiles\053d2a5848ae9d6f48c061efd95af32d\slang-core-module-generated.h.rule' exited with code 9009.
21>Done building project "slang-embedded-core-module.vcxproj" -- FAILED.
22>------ Build started: Project: PK_Gen, Configuration: Debug x64 ------
22>Scanning sources for module dependencies...
22>main.obj : error LNK2001: unresolved external symbol __imp_spReflection_ToJson
22>main.obj : error LNK2001: unresolved external symbol __imp_slang_createGlobalSession
22>D:\DEV\Github\PK_Gen\.PROJ\Debug\bin\PK_Gen.exe : fatal error LNK1120: 2 unresolved externals
22>Done building project "PK_Gen.vcxproj" -- FAILED.
```

I tried to workaround that by calling `slangc` instead of `slang-bootstrap` by modifying `source/slang-core-module/CMakeLists.txt:19` from this:
```CMake
add_custom_command(
    OUTPUT ${core_module_generated_header}
    COMMAND
        slang-bootstrap -archive-type riff-lz4 -save-core-module-bin-source
        ${core_module_generated_header}
    DEPENDS slang-bootstrap
    VERBATIM
)
```
to this:
```CMake
add_custom_command(
    OUTPUT ${core_module_generated_header}
    COMMAND
        slangc -archive-type riff-lz4 -save-stdlib-bin-source
        ${core_module_generated_header}
    VERBATIM
)
```

But since I removed `slang-bootstrap` from dependencies, my own app had unresolved symbols (you can see it at the end of the error log).
I tried to workaround that by linking to `slang-common-objects`, it seemed like working, but when I run my app, it just stalled forever.
It turns out `createGlobalSession()` fails to load embedded core module binary at `source/slang/slang-api.cpp:103`. I don't think binary would be incompatible and `slang-bootstrap` was just renamed `slangc`, but since `-save-core-module-bin-source` was incompatible with `slangc`, I think bootstrap was custom :D

At the end, I couldn't get a workaround working, but I hope my little research helps. I could probably fix it and create a pull request, if I was more familiar with the project and slang's custom cmake stuff

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -385,12 +385,14 @@ configure_package_config_file(
 # linkable targets. In this case do not export the targets. Otherwise, just
 # export the slang targets. 
 if(NOT CMAKE_SYSTEM_NAME STREQUAL "Emscripten")
-    install(
-        EXPORT SlangTargets
-        FILE ${PROJECT_NAME}Targets.cmake
-        NAMESPACE ${PROJECT_NAME}::
-        DESTINATION cmake
-    )
+    if(NOT ${SLANG_BUILD_TYPE} STREQUAL "STATIC")
+        install(
+            EXPORT SlangTargets
+            FILE ${PROJECT_NAME}Targets.cmake
+            NAMESPACE ${PROJECT_NAME}::
+            DESTINATION cmake
+        )
+    endif()
 endif()
 
 install(

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -193,6 +193,8 @@ option(
     ON
 )
 
+option(SLANG_ENABLE_RELEASE_LTO "Enable LTO for Release builds" ON)
+
 option(
     SLANG_ENABLE_SPLIT_DEBUG_INFO
     "Generate split debug info for debug builds"

--- a/cmake/SlangTarget.cmake
+++ b/cmake/SlangTarget.cmake
@@ -166,12 +166,14 @@ function(slang_add_target dir type)
 
     # Enable link-time optimization for release builds
     # See: https://cmake.org/cmake/help/latest/prop_tgt/INTERPROCEDURAL_OPTIMIZATION.html
-    set_target_properties(
-        ${target}
-        PROPERTIES
-            INTERPROCEDURAL_OPTIMIZATION_RELEASE TRUE
-            INTERPROCEDURAL_OPTIMIZATION_RELWITHDEBINFO TRUE
-    )
+    if(SLANG_ENABLE_RELEASE_LTO)
+        set_target_properties(
+            ${target}
+            PROPERTIES
+                INTERPROCEDURAL_OPTIMIZATION_RELEASE TRUE
+                INTERPROCEDURAL_OPTIMIZATION_RELWITHDEBINFO TRUE
+        )
+    endif()
 
     #
     # Set the output directory

--- a/docs/building.md
+++ b/docs/building.md
@@ -165,6 +165,7 @@ See the [documentation on testing](../tools/slang-test/README.md) for more infor
 | `SLANG_ENABLE_EXAMPLES`           | `TRUE`                     | Enable example targets, requires SLANG_ENABLE_GFX                                            |
 | `SLANG_LIB_TYPE`                  | `SHARED`                   | How to build the slang library                                                               |
 | `SLANG_ENABLE_RELEASE_DEBUG_INFO` | `TRUE`                     | Enable generating debug info for Release configs                                             |
+| `SLANG_ENABLE_RELEASE_LTO`        | `TRUE`                     | Enable LTO for Release builds                                                                |
 | `SLANG_ENABLE_SPLIT_DEBUG_INFO`   | `TRUE`                     | Enable generating split debug info for Debug and RelWithDebInfo configs                      |
 | `SLANG_SLANG_LLVM_FLAVOR`         | `FETCH_BINARY_IF_POSSIBLE` | How to set up llvm support                                                                   |
 | `SLANG_SLANG_LLVM_BINARY_URL`     | System dependent           | URL specifying the location of the slang-llvm prebuilt library                               |

--- a/prelude/CMakeLists.txt
+++ b/prelude/CMakeLists.txt
@@ -23,12 +23,14 @@ endforeach()
 slang_add_target(
     .
     OBJECT
+    EXPORT_MACRO_PREFIX SLANG
     EXPLICIT_SOURCE ${prelude_source}
     EXCLUDE_FROM_ALL
     TARGET_NAME prelude
     INCLUDE_DIRECTORIES_PUBLIC
         ${CMAKE_CURRENT_LIST_DIR}
         ${CMAKE_CURRENT_LIST_DIR}/../include
+    EXPORT_TYPE_AS ${SLANG_LIB_TYPE}
     LINK_WITH_PRIVATE unordered_dense::unordered_dense
     PUBLIC_HEADERS ${CMAKE_CURRENT_LIST_DIR}/slang*.h
     # It's an object library, so the install step only installs the headers

--- a/source/compiler-core/CMakeLists.txt
+++ b/source/compiler-core/CMakeLists.txt
@@ -1,6 +1,7 @@
 slang_add_target(
     .
     STATIC
+    EXPORT_MACRO_PREFIX SLANG
     EXCLUDE_FROM_ALL
     USE_EXTRA_WARNINGS
     LINK_WITH_PRIVATE core

--- a/source/core/CMakeLists.txt
+++ b/source/core/CMakeLists.txt
@@ -1,6 +1,7 @@
 slang_add_target(
     .
     STATIC
+    EXPORT_MACRO_PREFIX SLANG
     EXCLUDE_FROM_ALL
     USE_EXTRA_WARNINGS
     LINK_WITH_PRIVATE miniz lz4_static Threads::Threads ${CMAKE_DL_LIBS}

--- a/source/slang-rt/CMakeLists.txt
+++ b/source/slang-rt/CMakeLists.txt
@@ -1,7 +1,7 @@
 if(SLANG_ENABLE_SLANGRT)
     slang_add_target(
         .
-        SHARED
+        ${SLANG_LIB_TYPE}
         # This compiles 'core' again with the SLANG_RT_DYNAMIC_EXPORT macro defined
         EXTRA_SOURCE_DIRS ${slang_SOURCE_DIR}/source/core
         USE_EXTRA_WARNINGS

--- a/source/slang/CMakeLists.txt
+++ b/source/slang/CMakeLists.txt
@@ -28,8 +28,10 @@ add_custom_command(
 slang_add_target(
     slang-capability-defs
     OBJECT
+    EXPORT_MACRO_PREFIX SLANG
     USE_EXTRA_WARNINGS
     EXPLICIT_SOURCE ${SLANG_CAPABILITY_GENERATED_HEADERS}
+    EXPORT_TYPE_AS ${SLANG_LIB_TYPE}
     LINK_WITH_PRIVATE core
     INCLUDE_DIRECTORIES_PUBLIC
         "${SLANG_CAPABILITY_OUTPUT_DIR}"
@@ -40,8 +42,10 @@ slang_add_target(
 slang_add_target(
     slang-capability-lookup
     OBJECT
+    EXPORT_MACRO_PREFIX SLANG
     USE_EXTRA_WARNINGS
     EXPLICIT_SOURCE ${SLANG_CAPABILITY_GENERATED_SOURCE}
+    EXPORT_TYPE_AS ${SLANG_LIB_TYPE}
     LINK_WITH_PRIVATE core slang-capability-defs
     EXCLUDE_FROM_ALL
     FOLDER generated
@@ -157,10 +161,12 @@ add_custom_command(
 slang_add_target(
     slang-lookup-tables
     OBJECT
+    EXPORT_MACRO_PREFIX SLANG
     USE_EXTRA_WARNINGS
     EXPLICIT_SOURCE
         ${SLANG_LOOKUP_GENERATED_SOURCE}
         ${SLANG_SPIRV_CORE_GRAMMAR_SOURCE}
+    EXPORT_TYPE_AS ${SLANG_LIB_TYPE}
     LINK_WITH_PRIVATE core SPIRV-Headers
     EXCLUDE_FROM_ALL
     FOLDER generated
@@ -274,6 +280,7 @@ else()
     slang_add_target(
         .
         ${SLANG_LIB_TYPE}
+        EXPORT_MACRO_PREFIX SLANG
         ${slang_link_args}
         ${slang_interface_args}
         NO_SOURCE
@@ -291,6 +298,7 @@ else()
     slang_add_target(
         .
         ${SLANG_LIB_TYPE}
+        EXPORT_MACRO_PREFIX SLANG
         ${slang_link_args}
         ${slang_interface_args}
         ${slang_public_lib_args}

--- a/tools/CMakeLists.txt
+++ b/tools/CMakeLists.txt
@@ -73,6 +73,7 @@ generator(
 slang_add_target(
     slang-cpp-parser
     STATIC
+    EXPORT_MACRO_PREFIX SLANG
     USE_FEWER_WARNINGS
     LINK_WITH_PRIVATE core compiler-core
     INCLUDE_DIRECTORIES_PUBLIC .
@@ -104,7 +105,7 @@ if(SLANG_ENABLE_GFX)
     #
     slang_add_target(
         platform
-        SHARED
+        ${SLANG_LIB_TYPE}
         EXCLUDE_FROM_ALL
         USE_FEWER_WARNINGS
         LINK_WITH_PRIVATE
@@ -128,7 +129,7 @@ if(SLANG_ENABLE_GFX)
     #
     slang_add_target(
         gfx
-        SHARED
+        ${SLANG_LIB_TYPE}
         USE_FEWER_WARNINGS
         LINK_WITH_PRIVATE
             core

